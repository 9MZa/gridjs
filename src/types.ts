import { ComponentChild } from 'preact';
import Row from './row';
import { SortConfig } from './view/plugin/sort/sort';

export type ProtoExtends<T, U> = U & Omit<T, keyof U>;

export type OneDArray<T> = T[];
export type TwoDArray<T> = T[][];

// Table cell types
export type TCell = number | string | boolean | ComponentChild | HTMLElement;
export type TData = TwoDArray<TCell> | OneDArray<{ [key: string]: TCell }>;

// Table header cell type
export interface TColumn {
  id?: string;
  name: string;
  width?: string;
  sort?: SortConfig;
  children?: OneDArray<TColumn>;
  formatter?: (cell: TCell, row: Row, column: TColumn) => ComponentChild;
}

// Comparator function for the sorting plugin
export type Comparator<T> = (a: T, b: T) => number;

export interface TColumnSort {
  index: number;
  // 1 ascending, -1 descending
  direction?: 1 | -1;
}

// container status
export enum Status {
  Init,
  Loading,
  Loaded,
  Rendered,
  Error,
}

export type CSSDeclaration = {
  [key: string]: string;
};
