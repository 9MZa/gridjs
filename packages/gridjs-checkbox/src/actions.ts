import { BaseActions } from 'gridjs';
import { ID } from 'gridjs';

export interface CheckboxActionsType {
  CHECK: {
    ROW_ID: ID;
  };

  UNCHECK: {
    ROW_ID: ID;
  };
}

export class CheckboxActions extends BaseActions<CheckboxActionsType> {
  check(rowId: ID): void {
    this.dispatch('CHECK', {
      ROW_ID: rowId,
    });
  }

  uncheck(rowId: ID): void {
    this.dispatch('UNCHECK', {
      ROW_ID: rowId,
    });
  }
}
