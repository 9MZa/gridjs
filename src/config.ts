import { OneDArray, ProtoExtends, TCell, TColumn, TwoDArray } from './types';
import Storage from './storage/storage';
import Pipeline from './pipeline/pipeline';
import Tabular from './tabular';
import { SearchConfig } from './view/plugin/search/search';
import { PaginationConfig } from './view/plugin/pagination';
import Header from './header';
import { ServerStorageOptions } from './storage/server';
import Dispatcher from './util/dispatcher';
import { GenericSortConfig } from './view/plugin/sort/sort';
import { Language, Translator } from './i18n/language';
import { createRef, RefObject } from 'preact';
import StorageUtils from './storage/storageUtils';
import PipelineUtils from './pipeline/pipelineUtils';

// Config type used internally
export interface Config {
  dispatcher?: Dispatcher<any>;
  /** container element that is used to mount the Grid.js to */
  container?: Element;
  /** gridjs-temp div which is used internally */
  tempRef?: RefObject<HTMLDivElement>;
  data?:
    | TwoDArray<TCell>
    | (() => TwoDArray<TCell>)
    | (() => Promise<TwoDArray<TCell>>);
  server?: ServerStorageOptions;
  header?: Header;
  /** to parse a HTML table and load the data */
  from: HTMLElement;
  storage: Storage<any>;
  pipeline: Pipeline<Tabular>;
  /** to automatically calculate the columns width */
  autoWidth: boolean;
  /** sets the width of the container and table */
  width: string;
  search: SearchConfig;
  pagination: PaginationConfig;
  sort: GenericSortConfig;
  translator: Translator;
}

// Config type used by the consumers
interface UserConfigExtend {
  columns?: OneDArray<TColumn | string>;
  search: SearchConfig | boolean;
  pagination: PaginationConfig | boolean;
  // implicit option to enable the sort plugin globally
  sort: GenericSortConfig | boolean;
  language: Language;
}

export type UserConfig = ProtoExtends<
  Partial<Config>,
  Partial<UserConfigExtend>
>;

export class Config {
  constructor(config?: Config) {
    // FIXME: not sure if this makes sense because Config is a subset of UserConfig
    const updatedConfig = {
      ...Config.defaultConfig(),
      ...config,
    };

    Object.assign(this, updatedConfig);
  }

  update(updatedConfig: Partial<Config>): Config {
    Object.assign(this, updatedConfig || {});
    return this;
  }

  static defaultConfig(): Config {
    return {
      tempRef: createRef(),
      width: '100%',
      autoWidth: true,
    } as Config;
  }

  static fromUserConfig(userConfig: UserConfig): Config {
    const config = new Config(userConfig as Config);

    config.update({
      dispatcher: new Dispatcher<any>(),
      storage: StorageUtils.createFromUserConfig(userConfig),
    });

    config.update({
      pipeline: PipelineUtils.createFromConfig(config),
    });

    // Sort
    if (typeof userConfig.sort === 'boolean' && userConfig.sort) {
      config.update({
        sort: {
          multiColumn: true,
        },
      });
    }

    // Header
    config.update({
      header: Header.fromUserConfig(config),
    });

    // Pagination
    config.update({
      pagination: {
        enabled:
          userConfig.pagination === true ||
          userConfig.pagination instanceof Object,
        ...(userConfig.pagination as PaginationConfig),
      },
    });

    // Search
    config.update({
      search: {
        enabled:
          userConfig.search === true || userConfig.search instanceof Object,
        ...(userConfig.search as SearchConfig),
      },
    });

    // Translator
    config.update({
      translator: new Translator(userConfig.language),
    });

    return config;
  }
}
