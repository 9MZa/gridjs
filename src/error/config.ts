import BaseError from './base';

class ConfigError extends BaseError {}
export default ConfigError;
