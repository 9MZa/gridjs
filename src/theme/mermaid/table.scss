table.gridjs {
  &-table {
    border-collapse: collapse;
    text-align: left;
    display: table;
    margin: 0;
    padding: 0;
    overflow: auto;
  }
}
