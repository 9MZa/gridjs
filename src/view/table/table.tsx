import { h } from 'preact';

import Tabular from '../../tabular';
import { TBody } from './tbody';
import { THead } from './thead';
import { BaseComponent, BaseProps } from '../base';
import Header from '../../header';
import { className } from '../../util/className';
import { Status, TCell } from '../../types';

interface TableProps extends BaseProps {
  data: Tabular<TCell>;
  status: Status;
  header?: Header;
  width?: string;
}

export class Table extends BaseComponent<TableProps, {}> {
  private getStyle(): { [key: string]: string } {
    const style = {};

    if (this.props.width) {
      style['width'] = this.props.width;
    }

    return style;
  }

  render() {
    return (
      <table className={className('table')} style={this.getStyle()}>
        <THead header={this.props.header} />
        <TBody
          data={this.props.data}
          status={this.props.status}
          header={this.props.header}
        />
      </table>
    );
  }
}
