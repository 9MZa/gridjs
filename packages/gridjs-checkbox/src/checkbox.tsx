import { h } from 'gridjs';
import { CheckboxStore, CheckboxStoreState } from './store';
import { CheckboxActions } from './actions';
import { className } from 'gridjs';
import { Row } from 'gridjs';
import { PluginBaseComponent, PluginBaseProps } from 'gridjs';
import { Cell } from 'gridjs';

interface CheckboxState {
  isChecked: boolean;
}

interface CheckboxProps {
  // it's optional because thead doesn't have a row
  row?: Row;
  cell?: Cell;
  checkboxStore?: CheckboxStore;
  selectedClassName?: string;
  checkboxClassName?: string;
}

export class Checkbox extends PluginBaseComponent<
  CheckboxProps & PluginBaseProps<Checkbox>,
  CheckboxState
> {
  private readonly actions: CheckboxActions;
  private readonly store: CheckboxStore;
  private readonly storeUpdatedFn: (...args) => void;

  private isDataCell = (props): boolean => props.row !== undefined;
  private getParentTR = (): Element =>
    this.base &&
    this.base.parentElement &&
    (this.base.parentElement.parentElement as Element);

  static defaultProps = {
    selectedClassName: className('tr', 'selected'),
    checkboxClassName: className('checkbox'),
  };

  constructor(props: CheckboxProps & PluginBaseProps<Checkbox>, context) {
    super(props, context);

    this.state = {
      isChecked: false,
    };

    // store/dispatcher is required only if we are rendering a TD (not a TH)
    if (this.isDataCell(props)) {
      // create a new store if a global store doesn't exist
      if (!props.checkboxStore) {
        const store = new CheckboxStore(this.config.dispatcher);
        this.store = store;

        // to reuse for other checkboxes
        props.plugin.props.checkboxStore = store;
      } else {
        // restore the existing store
        this.store = props.checkboxStore;
      }

      this.actions = new CheckboxActions(this.config.dispatcher);
      this.storeUpdatedFn = this.storeUpdated.bind(this);
      this.store.on('updated', this.storeUpdatedFn);

      // also mark this checkbox as checked if cell.data is true
      if (props.cell.data) {
        this.check();
      }
    }
  }

  componentWillUnmount(): void {
    this.store.off('updated', this.storeUpdatedFn);
  }

  componentDidMount(): void {
    if (this.store) this.storeUpdated(this.store.state);
  }

  private storeUpdated(state: CheckboxStoreState): void {
    const parent = this.getParentTR();

    if (!parent) return;

    const isChecked = state.rowIds.indexOf(this.props.row.id) > -1;

    this.setState({
      isChecked: isChecked,
    });

    if (isChecked) {
      parent.classList.add(this.props.selectedClassName);
    } else {
      parent.classList.remove(this.props.selectedClassName);
    }
  }

  private check(): void {
    this.actions.check(this.props.row.id);
    this.props.cell.update(true);
  }

  private uncheck(): void {
    this.actions.uncheck(this.props.row.id);
    this.props.cell.update(false);
  }

  private toggle(): void {
    if (this.state.isChecked) {
      this.uncheck();
    } else {
      this.check();
    }
  }

  render() {
    if (this.isDataCell(this.props)) {
      return (
        <input
          type={'checkbox'}
          checked={this.state.isChecked}
          onChange={() => this.toggle()}
          className={this.props.checkboxClassName}
        />
      );
    }

    return null;
  }
}
