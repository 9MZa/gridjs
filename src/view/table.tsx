import { h } from 'preact';

import Tabular from '../tabular';
import { TBody } from './tbody';
import { THead } from './thead';
import { BaseComponent, BaseProps } from './base';
import Header from '../header';
import className from '../util/className';
import Config from '../config';
import { TBodyCell } from '../types';

import '../theme/mermaid/table.scss';

interface TableProps extends BaseProps {
  data?: Tabular<TBodyCell>;
  header?: Header;
}

export class Table extends BaseComponent<TableProps, {}> {
  render() {
    return (
      <table
        className={className(Config.current.classNamePrefix, 'table')}
        style={{ width: Config.current.width }}
      >
        <THead header={this.props.header} />
        <TBody data={this.props.data} />
      </table>
    );
  }
}
