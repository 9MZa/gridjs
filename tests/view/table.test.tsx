import { mount } from 'enzyme';
import { h } from 'preact';
import { Table } from '../../src/view/table/table';
import Header from '../../src/header';
import { Config } from '../../src/config';
import StorageUtils from '../../src/storage/storageUtils';
import Pipeline from '../../src/pipeline/pipeline';
import StorageExtractor from '../../src/pipeline/extractor/storage';
import ArrayToTabularTransformer from '../../src/pipeline/transformer/arrayToTabular';
import { Status } from '../../src/types';

describe('Table component', () => {
  let config: Config;

  beforeAll(() => {
    config = new Config();
    config.data = [
      [1, 2, 3],
      ['a', 'b', 'c'],
    ];

    config.storage = StorageUtils.createFromUserConfig(config);
    config.pipeline = new Pipeline([
      new StorageExtractor({ storage: config.storage }),
      new ArrayToTabularTransformer(),
    ]);
  });

  it('should render a table', async () => {
    const table = mount(
      <Table
        pipeline={config.pipeline}
        data={await config.pipeline.process()}
        status={Status.Loaded}
      />,
    );

    expect(table.html()).toMatchSnapshot();
  });

  it('should render a table with loading', async () => {
    const table = mount(
      <Table
        pipeline={config.pipeline}
        data={await config.pipeline.process()}
        status={Status.Loading}
      />,
    );

    expect(table.html()).toMatchSnapshot();
  });

  it('should render a table with header', async () => {
    const table = mount(
      <Table
        pipeline={config.pipeline}
        data={await config.pipeline.process()}
        header={Header.fromUserConfig({ columns: ['h1', 'h2', 'h3'] })}
        status={Status.Loaded}
      />,
    );

    expect(table.html()).toMatchSnapshot();
  });

  it('should render a table with width', async () => {
    const table = mount(
      <Table
        pipeline={config.pipeline}
        data={await config.pipeline.process()}
        width="300px"
        header={Header.fromUserConfig({ columns: ['h1', 'h2', 'h3'] })}
        status={Status.Loaded}
      />,
    );

    expect(table.html()).toMatchSnapshot();
  });

  it('should render a table with column width', async () => {
    const header = Header.fromUserConfig({ columns: ['h1', 'h2', 'h3'] });
    header.columns[0].width = '10%';
    header.columns[2].width = '300px';

    const table = mount(
      <Table
        pipeline={config.pipeline}
        data={await config.pipeline.process()}
        header={header}
        status={Status.Loaded}
      />,
    );

    expect(table.html()).toMatchSnapshot();
  });

  it('should render a table with column sort', async () => {
    const header = Header.fromUserConfig({ columns: ['h1', 'h2', 'h3'] });
    header.columns[0].sort = {
      enabled: true,
    };
    header.columns[2].sort = {
      enabled: true,
    };

    const table = mount(
      <Table
        pipeline={config.pipeline}
        data={await config.pipeline.process()}
        header={header}
        status={Status.Loaded}
      />,
    );

    expect(table.html()).toMatchSnapshot();
  });
});
