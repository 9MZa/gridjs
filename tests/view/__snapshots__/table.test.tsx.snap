// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table component should render a table 1`] = `"<table class=\\"gridjs gridjs-table\\"><tbody class=\\"gridjs gridjs-tbody\\"><tr class=\\"gridjs gridjs-tr\\"><td class=\\"gridjs gridjs-td\\">1</td><td class=\\"gridjs gridjs-td\\">2</td><td class=\\"gridjs gridjs-td\\">3</td></tr><tr class=\\"gridjs gridjs-tr\\"><td class=\\"gridjs gridjs-td\\">a</td><td class=\\"gridjs gridjs-td\\">b</td><td class=\\"gridjs gridjs-td\\">c</td></tr></tbody></table>"`;

exports[`Table component should render a table with header 1`] = `"<table class=\\"gridjs gridjs-table\\"><thead class=\\"gridjs gridjs-thead\\"><tr class=\\"gridjs gridjs-tr\\"><th class=\\"gridjs gridjs-th\\">h1</th><th class=\\"gridjs gridjs-th\\">h2</th><th class=\\"gridjs gridjs-th\\">h3</th></tr></thead><tbody class=\\"gridjs gridjs-tbody\\"><tr class=\\"gridjs gridjs-tr\\"><td class=\\"gridjs gridjs-td\\">1</td><td class=\\"gridjs gridjs-td\\">2</td><td class=\\"gridjs gridjs-td\\">3</td></tr><tr class=\\"gridjs gridjs-tr\\"><td class=\\"gridjs gridjs-td\\">a</td><td class=\\"gridjs gridjs-td\\">b</td><td class=\\"gridjs gridjs-td\\">c</td></tr></tbody></table>"`;

exports[`Table component should render a table with width 1`] = `"<table class=\\"gridjs gridjs-table\\" style=\\"width: 300px;\\"><thead class=\\"gridjs gridjs-thead\\"><tr class=\\"gridjs gridjs-tr\\"><th class=\\"gridjs gridjs-th\\">h1</th><th class=\\"gridjs gridjs-th\\">h2</th><th class=\\"gridjs gridjs-th\\">h3</th></tr></thead><tbody class=\\"gridjs gridjs-tbody\\"><tr class=\\"gridjs gridjs-tr\\"><td class=\\"gridjs gridjs-td\\">1</td><td class=\\"gridjs gridjs-td\\">2</td><td class=\\"gridjs gridjs-td\\">3</td></tr><tr class=\\"gridjs gridjs-tr\\"><td class=\\"gridjs gridjs-td\\">a</td><td class=\\"gridjs gridjs-td\\">b</td><td class=\\"gridjs gridjs-td\\">c</td></tr></tbody></table>"`;
